File: PythonSolutions/Neetcode150/binarytree_problems/98_validate-bst.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        # Preorder traversal, uses params to know lowest and highest in current path
        def validate(node, low, high):
            if not node:
                return True
            
            # Fail if current node is not in the range of lowest and highest in path
            if not (low < node.val < high):
                return False
            
            # Going left -> Pass current node value as new high
            # Going right -> Pass current node value as new low
            return validate(node.left, low, node.val) and validate(node.right, node.val, high)
        
        # Lowest and highest in the currernt path at the beginning are infinites
        return validate(root, float("-inf"), float("inf"))

File: TypescriptSolutions/Neetcode150/binary-tree_problems/98_validate-bst.ts
export {};
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}


const isValidBST = (root: TreeNode | null): boolean => {
    if (!root) return true;

    const validate = (node, low, high): boolean => {
        if (!node) return true;

        if (!(low < node.val && high > node.val)) {
            return false;
        }

        return validate(node.left, low, node.val) && validate(node.right, node.val, high);
    }

    return validate(root, -Infinity, Infinity);
}; 